File: _BENCHMARK_SIM/code/poisson.py
#------------------------------------------------------------------------------#
# Created: 2026-01-08
# ~: OLS benchmarks for python functions
#------------------------------------------------------------------------------#

# NOTA: on 2026-01-08 pyfixest could not be installed on python 3.14 because 
#       of an incompatible scipy version.
#       One may need to use python <= 3.13 for pyfixest to be installed properly.
#       

# Quadruple loop over:
# 1) number of observations
# 2) number of repetitions
# 3) type of Data Generating Process (dgp)
# 4) number of fixed effects
# 

import time
import pyfixest
import pandas
import math

all_results_list = list()
for n_obs in (1e4, 1e5, 1e6):
  print(f"n = 1e{int(math.log10(n_obs))}")
  
  for id_rep in (1, 2, 3):
    print(f" rep = {id_rep}")
    
    # loading the data 
    filename = f"./data/data_n-1e{int(math.log10(n_obs))}_rep-{id_rep}.gz"
    base = pandas.read_csv(filename, compression='gzip')
    
    # we exponentiate the dep.var
    base['y'] = [math.exp(x) for x in base['y']]
      
    for dgp in ("simple", "difficult"):
      print(f"  dgp = {dgp}", flush = True, end = "")
      
      for n_fe in (2, 3):
        print(".", flush = True, end = "")
        
        # setting the formula depending on the cases
        if dgp == "simple":
          fml = "y ~ x1 + x2 | indiv_id + firm_id" if n_fe == 2 else "y ~ x1 + x2 | indiv_id + firm_id + year"
        elif dgp == "difficult":
          # we don't run for large obs bc run time > 1h
          if n_obs >= 1e6: 
            continue
          fml = "y ~ x1 + x2 | indiv_id + firm_id_difficult" if n_fe == 2 else "y ~ x1 + x2 | indiv_id + firm_id_difficult + year"
        
        
        #
        # pyfixest.fepois
        #
        
        time_start = time.perf_counter()
        est = pyfixest.fepois(fml, base)
        time_fepois = time.perf_counter() - time_start
        
        #
        # saving
        #
        
        result = pandas.DataFrame({
          'n': [n_obs], 'id_rep': [id_rep], 'dgp': [dgp], 'n_fe': [n_fe],
          'method': ["pyfixest.fepois"], 'time': [time_fepois]
        })
        
        all_results_list.append(result)
      
      print("")
    

# saving
all_results = pandas.concat(all_results_list)

all_results.to_csv("./results/results_poisson_py.csv", index = False)

File: _BENCHMARK_SIM/code/poisson.jl
# If you are running Julia for the first time, please read Readme.md
# 

# estimation
using CSV
using DataFrames
using GLFixedEffectModels, GLM, Distributions

all_results = DataFrame(n = Int[], id_rep = Int[], dgp = String[], 
                        n_fe = Int[], method = String[], time = Float64[])

for n_obs in [1e4, 1e5, 1e6]
  println("n = 1e$(Int(log10(n_obs)))")
  
  for id_rep in [1, 2, 3]
    println(" rep = $(id_rep)")
    
    # loading the data 
    filename = "./data/data_n-1e$(Int(log10(n_obs)))_rep-$id_rep.gz";
    base = CSV.read(filename, DataFrame);
    base.y = exp.(base.y);
    
    for dgp in ["simple", "difficult"]
      print("  dgp = $(dgp)")
    
      for n_fe in [2, 3]
        print(".")
        
        # setting up the formula
        if dgp == "simple"
          if n_fe == 2
            fml = @formula(y ~ x1 + x2 + fe(indiv_id) + fe(firm_id))
          else
            fml = @formula(y ~ x1 + x2 + fe(indiv_id) + fe(firm_id) + fe(year))
          end
        elseif dgp == "difficult"
          if n_fe == 2
            fml = @formula(y ~ x1 + x2 + fe(indiv_id) + fe(firm_id_difficult))
          else
            fml = @formula(y ~ x1 + x2 + fe(indiv_id) + fe(firm_id_difficult) + fe(year))
          end
        end
        
        #
        # GLFixedEffectModels.nlreg
        #
        
        time_start = time()
        est = nlreg(base, fml, Poisson(), LogLink());
        time_nlreg = time() - time_start
        
        
        #
        # saving
        #
        
        push!(all_results, (n_obs, id_rep, dgp, n_fe, "GLFixedEffectModels.nlreg", time_nlreg))
        
      end
      println("")
    end
  end
end


CSV.write("results/results_poisson_jl.csv", all_results)





